src/components/selection-list/selection-list-model.tsx

```diff
@@ -0,0 +1,158 @@
+            result[key] = data[schema[key]];
+        }
+    }
+    return result;
+}
+
+function defaultRenderItem(item: DataSourceItemDefaultFormat): React.ReactElement<any> | null {
+    if (typeof item === 'string') {
+        item = {value: item, label: item};
+    }
+
+    if (item === divider) {
+        return <Divider />;
+    } else if (item.hidden) {
+        return null;
+    }
+    return <Option value={item.value} disabled={item.disabled}>{item.label}</Option>;
+}
+
+export class SelectionListModel {
+    public items: SelectionListItem[] = [];
+
+    private selectableValues: SelectionListItemValue[] = [];
+    private selectedValue: SelectionListItemValue | undefined = undefined;
+    private focusedValue: SelectionListItemValue | undefined = undefined;
+
+    public addDataSource({dataSource = [], dataSchema, renderItem = defaultRenderItem}: OptionList) {
+        dataSource.forEach(data => {
+            const element = renderItem(dataSchema && typeof data === 'object' ? renameKeys(data, dataSchema) : data);
+            if (element) {
+                this.addItem(data, element);
+            }
+        });
+    }
+
+    public addChildren(children: React.ReactNode) {
+        React.Children.forEach(children, element => {
+            if (typeof element === 'object') {
+                this.addItem(element, element);
+            }
+        });
+    }
+
+    public focusFirst() {
+        this.focusIndex(0);
+    }
+
+    public focusLast() {
+        this.focusIndex(Infinity);
+    }
+
+    public focusPrevious() {
+        const currentIndex = this.getFocusedIndex();
+        this.focusIndex(currentIndex === -1 ? Infinity : currentIndex - 1);
+    }
+
+    public focusNext() {
+        const currentIndex = this.getFocusedIndex();
+        this.focusIndex(currentIndex === -1 ? 0 : currentIndex + 1);
+    }
+
+    public findItemByValue(value: SelectionListItemValue): SelectionListItem | undefined {
+        return this.items.find(item => item.value === value);
+    }
+
+    public getSelectedValue(): SelectionListItemValue | undefined {
+        return this.selectedValue;
+    }
+
+    public getFocusedValue(): SelectionListItemValue | undefined {
+        return this.focusedValue;
+    }
+
+    public selectValue(value: SelectionListItemValue | undefined) {
+        this.selectedValue = value;
+        for (const item of this.items) {
+            item.selected = (value !== undefined && item.value === value);
+        }
+    }
+
+    public focusValue(value: SelectionListItemValue | undefined) {
+        this.focusedValue = value;
+        for (const item of this.items) {
+            item.focused = (value !== undefined && item.value === value);
+        }
+    }
+
+    private addItem(data: SelectionListItemData, element: React.ReactElement<any>) {
+        const value = element.props.value;
+        const disabled = Boolean(element.props.disabled);
+        const item = extendShallowObservable(
```

Comment:
It turns out that having an observable for each item is faster both for updates (10 times faster than using expr) and for the initial render (by 10%).